### DISCUSSION TITLE
Adding full BSON support

### DISCUSSION THREAD

**Discussion by bazile-clyde** [nlohmann/json#2316] [Ideas]
@nlohmann, I'm interested in using this library to improve the usability of [MongoDB's C++ Driver](https://github.com/mongodb/mongo-cxx-driver). This would require support for _all_ [BSON types](http://bsonspec.org/spec.html). Currently, it looks like that'd mean adding support for the following types (let me know if I've missed anything):

- ObjectId (x07)	
- UTC datetime (x09) 	
- Regular expression (x0B)
- JavaScript code (x0D)
- Timestamp (x11)	
- 128-bit decimal floating-point (x13)
- Min key (xFF)	
- Max key (x7F)

I'm scoping out the amount of work I'll need to do and it seems like each type would require about as much effort as adding support for the binary type ([here](https://github.com/nlohmann/json/pull/1662)).

Are there any major issues I should know about before embarking on this journey?

**Comment by nlohmann** [comment]
I am afraid this library cannot provide "full" BSON support, because the library is centered around JSON which does not know all the types you mention. If we would add them to `basic_json`, then we would not only need to think about how to map them back to JSON text, but also to the other supported formats (CBOR, MessagePack, and UBJSON).

What I could think of would be parsing these types into a binary value using the type byte as subtype.

What do you think? Would this already be an improvement compared to the status quo.

**Reply by bazile-clyde** [reply]
Thanks for the speedy response. That makes sense. Unfortunately, I don't think this approach (representing the unsupported types as the byte subtype) is that much of an improvement over what we have now. Though, I'll reach out again if things change.

**Reply by nlohmann** [reply]
What kind of C++ types would you use to represent these types? I think this is quite specific to your application and hence wouldn't be something I would like to set in stone as library.

**Reply by bazile-clyde** [reply]
Currently, we use the C++ types listed below to represent all of the BSON types above in the C++ driver. I gave a (hopefully descriptive) name to the C++ types that are split between two values, e.g., regular expressions.

- ObjectId = `array<char, size_t>`
- UTC datetime = `chrono::milliseconds`
- Regular expression = `string regex, string options`
- JavaScript code = `string`
- Timestamp = `uint32_t timestamp, uint32_t incrementing_ordinal`
- 128-bit decimal floating-point = `uint64_t high, uint64_t low`
- Min key = `uint8_t`	
- Max key = `uint8_t`

We display these types to users using [extended JSON](https://docs.mongodb.com/master/reference/mongodb-extended-json/#bson.Timestamp).

**Reply by bazile-clyde** [reply]
We, of course, already have constructors for creating BSON (either with extended JSON or one of our builders [here](http://mongocxx.org/mongocxx-v3/working-with-bson/#builders)). So the big appeal of this library was the (in my opinion much simpler) JSON-like constructor.
